### PR TITLE
Add Parcelable implementations

### LINKED ISSUE #49: Parcelable implementation for Deferred?
**TLDR:** it'd be super useful for DeferredResources to extend `Parcelable`.

Here's my setup: 
- My ViewModels use a `val navigation: StateFlow<NavCommand>` to communicate with the UI layer, so that my Fragments don't have to concern themselves with the _what_ of navigation, just the _how_. 
- `NavCommand` is just a sealed class that wraps the different sort of operations one may take using the Nav component
```kotlin 
sealed class NavCommand {
    class To(val directions: NavDirections) : NavCommand()
    class Deeplink(...) : NavCommand()
    object Back : NavCommand()
    class BackTo(@IdRes val destinationId: Int, val inclusive: Boolean = false) : NavCommand()
}
```
- when the ViewModel decides navigation to another fragment needs to happen, typically it calls `navigation.value = To(SomethingFragmentDirections.actionFooToBar(a, b, c))`, where the inputs are the standard types that are acceptable in Bundles; one of those types is Parcelable. 
- at the UI layer, I say `val args by navArgs<NewFragmentArgs>()` (i.e. I use SafeArgs)

Because the Directions are instantiated at the VM-level, I can't `.resolve` the deferred resource at the time I (implicitly) build the Bundle.

Since I use safeargs, so as long as each concrete type knows how to serialize/deserialize from `Parcel`, I _think_ SafeArgs will code-gen the correct thing (i.e `DeferredText.Constant` won't come out the other side as `DeferredText.Resource`). 

Unfortunately, for people who _don't_ use SafeArgs, they'll have to manually remember to cast as `Constant` instead of `Resource`, otherwise they're gonna have a bad time.

### PR DESCRIPTION
**Pull request by drewhamilton**
Closes #49.

Adds `ParcelableDeferred*` interfaces and updates all concrete implementations to implement these interfaces. Each implementation is tested via a Bundle.

To-do before merging:
* ~~Create an issue for using the SafeArgs plugin in `androidTest/`~~ **Removed: SafeArgs uses normal Bundle/Parcelable APIs internally and we're already testing those.**
- [x] Existing concrete wrapper-like types, e.g. `FormattedDeferredText`, already accept non-Parcelable parameters. Is it a good idea to make these classes Parcelable, with the possibility that they will fail at runtime if someone constructs them with a non-Parcelable wrapped member and then tries to pass them through a Bundle? Should I overload the constructors to provide some kind of deprecation or other warning? **Decided: I will leave this potential runtime failure as is, with the expectation that consumers using the Parcelable functionality will also use Parcelable params. If it poses problems consumers can file issues and then we can consider a lint warning or something.**
- [x] Existing extension function return types can't be changed and this is a problem. I want to solve this with Overload Returns but it's [not working](https://github.com/JakeWharton/OverloadReturn/issues/2). What to do? **Solved: using deprecation level HIDDEN and adding an unused generic to the existing extensions: 224142fb280bd11b36f5f60e9d3dfe127a3464f6**

### CODE REVIEW

**Comment by drewhamilton** [comment]
@JvmName I did DeferredText first, and for most of the other types I think it will be equally straightforward. Typeface and Drawable are the outliers because those non-deferred types aren't Parcelable themselves. I'm thinking I'll just ignore them unless you have a better idea.

Let me know if this covers your use case. Also feel free to review, and let me know if you want to add to the code!

**Comment by JvmName** [comment]
This is looking cool! Thanks for such a quick turnaround 😄

**Comment by drewhamilton** [comment]
Some more things to consider before eventually merging this:

- [ ] Existing concrete wrapper-like types, e.g. `FormattedDeferredText`, already accept non-Parcelable parameters. Is it a good idea to make these classes Parcelable, with the possibility that they will fail at runtime if someone constructs them with a non-Parcelable wrapped member and then tries to pass them through a Bundle? Should I overload the constructors to provide some kind of deprecation or other warning?
- [ ] Existing extension function return types can't be changed and this is a problem. I want to solve this with Overload Returns but it's [not working](https://github.com/JakeWharton/OverloadReturn/issues/2). What to do?

File: deferred-resources/src/main/java/com/backbase/deferredresources/text/FormattedDeferredText.kt [hunk 1]
```diff
@@ -4,26 +4,39 @@ import android.content.Context
...
  */
-@DataApi public class FormattedDeferredText private constructor(
-    // Private constructor marker allows vararg constructor overload while retaining DataApi toString generation
-    @Suppress("UNUSED_PARAMETER") privateConstructorMarker: Int,
-    private val wrapped: DeferredFormattedString,
-    private val formatArgs: Array<out Any>
-) : DeferredText {
+// Primary constructor is internal rather than private so the generated Creator can access it
+@Parcelize
+@DataApi public class FormattedDeferredText internal constructor(
```
Comment (JvmName) [inline deferred-resources/src/main/java/com/backbase/deferredresources/text/FormattedDeferredText.kt:36]: will this cause issues if someone is using a newer version of Poko/ExtraCare?

**Comment by JvmName** [comment]
This looks awesome, looking forward to using it 🎉 

Thanks for all the hard work 🙏🏾

File: deferred-resources/src/main/java/com/backbase/deferredresources/text/FormattedDeferredText.kt (hunk 1 above)
Comment (drewhamilton) [reply deferred-resources/src/main/java/com/backbase/deferredresources/text/FormattedDeferredText.kt:36]: No, it shouldn't – it's applied at compile time of this library, so the generated methods are already in the binary by the time it reaches the consuming project, and the consuming project is free to use whichever versions of Kotlin and/or Poko it wants.

**Pull request merged**